### PR TITLE
Add GCS dependencies to resolve_dependency and refactor cmake modules

### PR DESCRIPTION
**Pull request by tigrux**
The goal of this change is to refactor and simplify the existing cmake modules.

### CODE REVIEW

File: CMake/Findre2.cmake
Comment (assignUser) [inline CMake/Findre2.cmake:1]: Why is this necessary?

File: CMake/resolve_dependency_modules/absl.cmake [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+FetchContent_Declare(
+  absl
+  URL ${VELOX_ABSL_SOURCE_URL}
+  URL_HASH ${VELOX_ABSL_BUILD_SHA256_CHECKSUM}
+  OVERRIDE_FIND_PACKAGE EXCLUDE_FROM_ALL)
+
+set(ABSL_BUILD_TESTING OFF)
+set(ABSL_PROPAGATE_CXX_STD ON)
+set(ABSL_ENABLE_INSTALL ON)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/absl.cmake:35]: Does abseil respect the install prefix and install into the binary dir? (I have seen libraries that install into hardcoded paths)

File: CMake/resolve_dependency_modules/protobuf.cmake [hunk 2]
```diff
@@ -30,44 +30,10 @@ message(STATUS "Building Protobuf from source")
...
-  else()
-    set(Protobuf_LIBRARIES "${protobuf_BINARY_DIR}/libprotobuf.a")
-  endif()
-  add_subdirectory(${protobuf_SOURCE_DIR} ${protobuf_BINARY_DIR})
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS_BKP}")
-endif()
+set(protobuf_BUILD_TESTS OFF)
+FetchContent_MakeAvailable(protobuf)
+set(Protobuf_INCLUDE_DIRS ${protobuf_SOURCE_DIR}/src)
+set(Protobuf_PROTOC_EXECUTABLE "${protobuf_BINARY_DIR}/protoc")
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/protobuf.cmake:39]: We should just use the proper targets instead of the vars but that's out of scope for this PR

File: CMakeLists.txt [hunk 3]
```diff
@@ -465,14 +468,17 @@ else()
...
-  set_source(Protobuf)
-  resolve_dependency(Protobuf 3.21 EXACT)
-  include_directories(${Protobuf_INCLUDE_DIRS})
+if(VELOX_ENABLE_GCS)
+  set_source(absl)
+  resolve_dependency(absl 20240116 EXACT)
+
+  set_source(gRPC)
+  resolve_dependency(gRPC 1.48.1 EXACT)
+
```
Comment (assignUser) [inline CMakeLists.txt:477]: You can move these into the gcp.cmake, they don't have to be called in the main cmake, that way it will only build them if gcp isn't found on the system (when it's found on the system all deps are also there).

File: scripts/setup-adapters.sh [hunk 4]
```diff
@@ -54,33 +54,26 @@ function install_aws_deps {
...
-  cmake_install -DBUILD_SHARED_LIBS=OFF \
-    -DABSL_BUILD_TESTING=OFF
-
   # crc32
   github_checkout google/crc32c 1.1.2 --depth 1
-  cmake_install -DBUILD_SHARED_LIBS=OFF \
+  cmake_install \
+    -DBUILD_SHARED_LIBS=OFF \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DBUILD_SHARED_LIBS=OFF \
```
Comment (assignUser) [inline scripts/setup-adapters.sh:65]: Why this addition, also build shared is duplicated.

File: scripts/setup-adapters.sh [hunk 5]
```diff
@@ -54,33 +54,26 @@ function install_aws_deps {
...
     -DCRC32C_BUILD_BENCHMARKS=OFF \
     -DCRC32C_USE_GLOG=OFF
 
   # nlohmann json
   github_checkout nlohmann/json v3.11.3 --depth 1
-  cmake_install -DBUILD_SHARED_LIBS=OFF \
+  cmake_install \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DBUILD_SHARED_LIBS=OFF \
+    -DBUILD_TESTING=OFF \
```
Comment (assignUser) [inline scripts/setup-adapters.sh:75]: Why the addition?

File: scripts/setup-adapters.sh [hunk 6]
```diff
@@ -54,33 +54,26 @@ function install_aws_deps {
...
     -DJSON_BuildTests=OFF
-
-  # google-cloud-cpp
-  github_checkout googleapis/google-cloud-cpp v2.22.0 --depth 1
-  cmake_install -DBUILD_SHARED_LIBS=OFF \
-    -DCMAKE_INSTALL_MESSAGE=NEVER \
-    -DGOOGLE_CLOUD_CPP_ENABLE_EXAMPLES=OFF \
-    -DGOOGLE_CLOUD_CPP_ENABLE=storage
```
Comment (assignUser) [inline scripts/setup-adapters.sh:83]: Why remove this? We want to install all deps into the docker image (or locally), even if they are in resolve_dependency.

File: scripts/setup-centos8.sh [hunk 7]
```diff
@@ -132,6 +140,38 @@ function install_protobuf {
...
+}
+
+function install_google_cloud_cpp {
+  wget_and_untar https://github.com/googleapis/google-cloud-cpp/archive/refs/tags/v2.24.0.tar.gz google_cloud_cpp
+  (
+    cd google_cloud_cpp
+    cmake_install -DBUILD_TESTING=OFF -DGOOGLE_CLOUD_CPP_ENABLE_EXAMPLES=OFF -DGOOGLE_CLOUD_CPP_ENABLE=storage,bigquery
+  )
+}
+
```
Comment (assignUser) [inline scripts/setup-centos8.sh:174]: Why is this here? it (and it's dependencies) should be in setup-adapters.sh

File: velox/dwio/common/CMakeLists.txt [hunk 8]
```diff
@@ -77,4 +75,5 @@ target_link_libraries(
   velox_memory
   Boost::regex
   Folly::folly
-  glog::glog)
+  glog::glog
+  protobuf::libprotobuf)
```
Comment (assignUser) [inline velox/dwio/common/CMakeLists.txt:79]: oh nice!

File: CMake/resolve_dependency_modules/protobuf.cmake (hunk 2 above)
Comment (assignUser) [reply CMake/resolve_dependency_modules/protobuf.cmake:39]: I see you did do that (I thought there were more targets that used it). So can we remove this? Or is gcp using the vars?

**Review by assignUser — COMMENTED** [review]
I see that it works with the current protobuf version, that's nice.

I think updating the cmake version is great `OVERRIDE_FINDE_PACKAGE` is really useful. I am not sure what we gain from `EXCLUDE_FROM_ALL` though? 

Any time resolve_dependencies get's called for any of these they will be needed as a dependency for other target's so they will get build?

File: CMake/Findre2.cmake
Comment (tigrux) [reply CMake/Findre2.cmake:1]: The regular Findre2.cmake did not work, so I copied the one shipped with gRPC, and this one worked.

File: CMake/resolve_dependency_modules/absl.cmake (hunk 1 above)
Comment (tigrux) [reply CMake/resolve_dependency_modules/absl.cmake:35]: Abseil is never installed, it it build as static libraries, so only .a files.

File: CMake/resolve_dependency_modules/protobuf.cmake (hunk 2 above)
Comment (tigrux) [reply CMake/resolve_dependency_modules/protobuf.cmake:39]: I tried to respect the existing .cmake files, to minimize the changes.

File: CMakeLists.txt (hunk 3 above)
Comment (tigrux) [reply CMakeLists.txt:477]: Yes, I agree.

File: scripts/setup-adapters.sh (hunk 4 above)
Comment (tigrux) [reply scripts/setup-adapters.sh:65]: Opsie! Too much copy-n-paste over the weekend.

File: scripts/setup-adapters.sh (hunk 5 above)
Comment (tigrux) [reply scripts/setup-adapters.sh:75]: Carelessly copy-n-paste from other scripts.

File: scripts/setup-adapters.sh (hunk 6 above)
Comment (tigrux) [reply scripts/setup-adapters.sh:83]: I was not aware that it was being used to create the image, I will add it back.

File: scripts/setup-centos8.sh (hunk 7 above)
Comment (tigrux) [reply scripts/setup-centos8.sh:174]: At the moment, it is the only option to have the CI build those libraries.

Comment (tigrux) [reply scripts/setup-centos8.sh:174]: My guess is that there is some magic trick to re-trigger a new docker image, but I do not know that trick yet.

File: CMake/resolve_dependency_modules/absl.cmake (hunk 1 above)
Comment (tigrux) [reply CMake/resolve_dependency_modules/absl.cmake:35]: The flag `ABSL_ENABLE_INSTALL` is needed for the targets to have the absl:: prefix namespace.

File: .github/workflows/linux-build.yml [hunk 9]
```diff
@@ -70,6 +70,16 @@ jobs:
         # it doesn't work
         run: git config --global --add safe.directory /__w/velox/velox
 
+      - name: Install cmake
+        run: python3 -m pip install cmake==3.28.3
```
Comment (bdice) [inline .github/workflows/linux-build.yml:74]: Suggestion: If it's possible to separate the CMake version bump into a smaller PR that can be reviewed/merged sooner, that would be helpful to me!

Comment (tigrux) [reply .github/workflows/linux-build.yml:74]: Of course! Precisely,  submitted this PR to collect this type of feedback.
I foresee at least 3 PR resulting from this one:
- Update to cmake 3.28.
- Refactor the existing cmake modules.
- Add the dependencies of Google Cloud CPP as cmake modules.

File: CMake/Findre2.cmake [hunk 10]
```diff
@@ -11,19 +11,55 @@
...
+find_package(PkgConfig REQUIRED)
+# TODO(junyer): Use the IMPORTED_TARGET option whenever CMake 3.6 (or newer)
+# becomes the minimum required: that will take care of the add_library() and
+# set_property() calls; then we can simply alias PkgConfig::RE2 as re2::re2. For
+# now, we can only set INTERFACE_* properties that existed in CMake 3.5.
+pkg_check_modules(RE2 QUIET re2)
+if(RE2_FOUND)
+  set(re2_FOUND "${RE2_FOUND}")
+  add_library(re2::re2 INTERFACE IMPORTED)
+  if(RE2_INCLUDE_DIRS)
```
Comment (kgpai) [inline CMake/Findre2.cmake:37]: curious why re2 has to do set al of these properties specifically ?

Comment (tigrux) [reply CMake/Findre2.cmake:37]: Hello, the old re2.cmake did not work with cmake 3.28. I noticed gRPC had its own re2.cmake module and I used it to replace Velox's one and it worked.

File: scripts/setup-centos8.sh [hunk 11]
```diff
@@ -50,8 +50,8 @@ function install_build_prerequisites {
   dnf_install epel-release dnf-plugins-core # For ccache, ninja
   dnf config-manager --set-enabled powertools
   dnf update -y
-  dnf_install ninja-build cmake curl ccache gcc-toolset-9 git wget which
-  dnf_install autoconf automake python39 python39-devel python39-pip libtool
+  dnf_install ninja-build curl ccache gcc-toolset-9 git wget which
+  dnf_install autoconf automake python39 python39-devel python39-pip libtool c-ares-devel
```
[the remark this answers is not in the document]
Comment (tigrux) [reply scripts/setup-centos8.sh:54]: c-ares is a dependency of Google Cloud C++.

File: scripts/setup-centos8.sh [hunk 12]
```diff
@@ -61,7 +61,7 @@ function install_velox_deps_from_dnf {
     libdwarf-devel curl-devel libicu-devel bison flex libsodium-devel
 
   # install sphinx for doc gen
-  pip3.9 install sphinx sphinx-tabs breathe sphinx_rtd_theme
+  pip3.9 install sphinx sphinx-tabs breathe sphinx_rtd_theme cmake==3.28.3
```
[the remark this answers is not in the document]
Comment (tigrux) [reply scripts/setup-centos8.sh:64]: I will fix it.

File: CMake/Findre2.cmake [hunk 13]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) Facebook, Inc. and its affiliates.
+# Copyright 2017 gRPC authors.
```
[the remark this answers is not in the document]
Comment (tigrux) [reply CMake/Findre2.cmake:1]: Yead, I copied it from gRPC repo.

File: scripts/setup-adapters.sh [hunk 14]
```diff
@@ -54,33 +54,51 @@ function install_aws_deps {
...
+  # Install gcs dependencies that cannot be installed via cmake's FetchContent
   # https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md#required-libraries
 
-  # abseil-cpp
-  github_checkout abseil/abseil-cpp 20240116.1 --depth 1
-  sed -i 's/^#define ABSL_OPTION_USE_\(.*\) 2/#define ABSL_OPTION_USE_\1 0/' "absl/base/options.h"
-  cmake_install -DBUILD_SHARED_LIBS=OFF \
-    -DABSL_BUILD_TESTING=OFF
+  # abseil
+  github_checkout abseil/abseil-cpp 20240116.2 --depth 1
```
[the remark this answers is not in the document]
Comment (tigrux) [reply scripts/setup-adapters.sh:61]: I have always used because it clones faster

--depth <depth>
           Create a shallow clone with a history truncated to the specified number of commits.

File: .github/workflows/scheduled.yml [hunk 15]
```diff
@@ -197,6 +197,7 @@ jobs:
         env:
           EXTRA_CMAKE_FLAGS: "-DVELOX_ENABLE_ARROW=ON -DVELOX_BUILD_PYTHON_PACKAGE=ON ${{ inputs.extraCMakeFlags }}"
         run: |
+          python3 -m pip install cmake==3.28.3
```
[the remark this answers is not in the document]
Comment (tigrux) [reply .github/workflows/scheduled.yml:200]: I don't have access to recreate the docker images, so I had to take extra steps to make the CI pass.

File: .github/workflows/linux-build.yml (hunk 9 above)
Comment (tigrux) [reply .github/workflows/linux-build.yml:74]: I don't know how to that from a regular PR.

File: CMake/Findre2.cmake (hunk 13 above)
[the remark this answers is not in the document]
Comment (assignUser) [reply CMake/Findre2.cmake:1]: We should start a NOTICE file to keep track of files like these similar to how arrow does it: https://github.com/apache/arrow/blob/main/NOTICE.txt

File: .github/workflows/scheduled.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (kgpai) [reply .github/workflows/scheduled.yml:200]: You can just change the setup/* scripts and publish them as a PR, if accepted and merged, the new docker images based on them will automatically be generated and available after merge to main.

File: .github/workflows/linux-build.yml (hunk 9 above)
Comment (kgpai) [reply .github/workflows/linux-build.yml:74]: Replied below for this, its pretty straightforward but does require another PR.

File: .github/workflows/scheduled.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (tigrux) [reply .github/workflows/scheduled.yml:200]: Excellent! Thanks @kgpai.

**Comment by assignUser** [comment]
Once #9861 is merged you can remove the install bits in the workflows and rebase on it (you'll have to wait a bit after the merge for the rebuild of the containers). I'll review after that but looks good so far, thanks for the contribution! :)

**Comment by tigrux** [comment]
Hello @majetideepak, @assignUser, @kgpai, @pedroerp, @bdice.
As agreed, here is the second part of the change to update to cmake 3.28.
This one uses the cmake support previously added to refactor the cmake modules.

File: CMake/resolve_dependency_modules/absl.cmake [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+resolve_dependency_url(ABSL)
+
+message(STATUS "Building Abseil from source")
+
+FetchContent_Declare(
+  absl
+  URL ${VELOX_ABSL_SOURCE_URL}
+  URL_HASH ${VELOX_ABSL_BUILD_SHA256_CHECKSUM}
+  OVERRIDE_FIND_PACKAGE EXCLUDE_FROM_ALL)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/absl.cmake:31]: Can you also add `SYSTEM`  (added in 3.25) this will suppress warnings.

File: CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+resolve_dependency_url(GOOGLE_CLOUD_CPP)
+
+message(STATUS "Building Google Cloud CPP storage from source")
+
+FetchContent_Declare(
+  google_cloud_cpp
+  URL ${VELOX_GOOGLE_CLOUD_CPP_SOURCE_URL}
+  URL_HASH ${VELOX_GOOGLE_CLOUD_CPP_BUILD_SHA256_CHECKSUM}
+  OVERRIDE_FIND_PACKAGE EXCLUDE_FROM_ALL)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake:38]: `SYSTEM`

File: CMake/resolve_dependency_modules/grpc.cmake [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+resolve_dependency_url(GRPC)
+
+message(STATUS "Building Google Cloud CPP from source")
+
+FetchContent_Declare(
+  gRPC
+  URL ${VELOX_GRPC_SOURCE_URL}
+  URL_HASH ${VELOX_GRPC_BUILD_SHA256_CHECKSUM}
+  OVERRIDE_FIND_PACKAGE EXCLUDE_FROM_ALL)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/grpc.cmake:31]: `SYSTEM`

File: CMake/resolve_dependency_modules/grpc.cmake [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+set(gRPC_PROTOBUF_PROVIDER
+    "package"
+    CACHE STRING "Provider of protobuf library")
+set(gRPC_INSTALL
+    ON
+    CACHE BOOL "Generate installation target")
+FetchContent_MakeAvailable(gRPC)
+add_library(gRPC::grpc ALIAS grpc)
+add_library(gRPC::grpc++ ALIAS grpc++)
+add_executable(gRPC::grpc_cpp_plugin ALIAS grpc_cpp_plugin)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/grpc.cmake:57]: :+1:

File: CMake/resolve_dependency_modules/protobuf.cmake [hunk 20]
```diff
@@ -30,44 +30,10 @@ message(STATUS "Building Protobuf from source")
 FetchContent_Declare(
   protobuf
   URL ${VELOX_PROTOBUF_SOURCE_URL}
-  URL_HASH ${VELOX_PROTOBUF_BUILD_SHA256_CHECKSUM})
+  URL_HASH ${VELOX_PROTOBUF_BUILD_SHA256_CHECKSUM}
+  OVERRIDE_FIND_PACKAGE EXCLUDE_FROM_ALL)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/protobuf.cmake:34]: SYSTEM

File: CMakeLists.txt [hunk 21]
```diff
@@ -442,6 +433,18 @@ if(${VELOX_BUILD_PYTHON_PACKAGE})
...
+# DWIO (ORC/DWRF) and Substrait depend on protobuf.
+if(${VELOX_BUILD_MINIMAL_WITH_DWIO}
+   OR ${VELOX_ENABLE_HIVE_CONNECTOR}
+   OR ${VELOX_ENABLE_SUBSTRAIT}
+   OR VELOX_ENABLE_GCS)
+
+  # Locate or build protobuf.
+  set_source(Protobuf)
+  resolve_dependency(Protobuf 3.21.4 EXACT)
+  include_directories(${Protobuf_INCLUDE_DIRS})
```
Comment (assignUser) [inline CMakeLists.txt:445]: I think we should be able to remove this due to the proper use of targets you added.

File: CMake/resolve_dependency_modules/protobuf.cmake
Comment (assignUser) [inline CMake/resolve_dependency_modules/protobuf.cmake:1]: What a great simplification!

File: CMake/resolve_dependency_modules/absl.cmake (hunk 16 above)
Comment (assignUser) [reply CMake/resolve_dependency_modules/absl.cmake:31]: Looking at the [protobuf build log ](https://github.com/facebookincubator/velox/actions/runs/9197613668/job/25298471631?pr=9072#step:7:1703) warnings don't stop the build because we set werror after adding the dependencies but I think it's still good to be explicit about this and add `SYSTEM` for consistency as imported targets (e.g. dependencies used from the host) also have this set.

**Review by assignUser — APPROVED** [review]
Looks good thanks! As the docker images have been updated you can remove the steps where you installed the recent cmake in #9861.

File: CMakeLists.txt (hunk 21 above)
Comment (tigrux) [reply CMakeLists.txt:445]: I tried but they are still needed.

Comment (assignUser) [reply CMakeLists.txt:445]: Ok we can review later on:)

**Review by assignUser — APPROVED** [review]
Once CI turns green this should be ready to merge.

**Review by bdice — APPROVED** [review]
This looks good from what I can tell.

File: CMake/Findre2.cmake (hunk 13 above)
[the remark this answers is not in the document]
Comment (kgpai) [reply CMake/Findre2.cmake:1]: @tigrux I believe we might have to revert this back.

[the remark this answers is not in the document]
Comment (tigrux) [reply CMake/Findre2.cmake]: Done!

**Review by kgpai — APPROVED** [review]
